library/ziti_enroll.c

```diff
@@ -91,210 +140,361 @@ int ziti_enroll(const ziti_enroll_opts *opts, uv_loop_t *loop,
+            er->opts.token = strdup(opts->token);
         }
+
+        rc = start_enrollment(er);
     }
 
-    return ERR(ziti);
+    if (rc != ZITI_OK) {
+        free_enroll_req(er);
+    }
+    return rc;
 }
 
-static void well_known_certs_cb(char *base64_encoded_pkcs7, const ziti_error *err, void *req) {
-    PREPF(ziti, ziti_errorstr);
+static void network_jwt_cb(ziti_network_jwt_array arr, const ziti_error *err, void *ctx) {
+    struct ziti_enroll_req *er = ctx;
+    if (err) {
+        er->enroll_cb(NULL, ZITI_ENROLLMENT_METHOD_UNSUPPORTED, ziti_errorstr(ZITI_ENROLLMENT_METHOD_UNSUPPORTED), er->ctx);
+        free_enroll_req(er);
+        return;
+    }
 
-    int ziti_err;
-    struct ziti_enroll_req *enroll_req = req;
-    if ((NULL == base64_encoded_pkcs7) || (NULL != err)) {
-        ZITI_LOG(DEBUG, "err->message is: %s", err->message);
-        TRY(ziti, ZITI_JWT_VERIFICATION_FAILED);
+    er->opts.token = strdup(arr[0]->token);
+    free_ziti_network_jwt_array(&arr);
+
+    int rc = start_enrollment(er);
+    if (rc != ZITI_OK) {
+        er->enroll_cb(NULL, rc, ziti_errorstr(rc), er->ctx);
+        free_enroll_req(er);
     }
+}
 
-    ZITI_LOG(VERBOSE, "base64_encoded_pkcs7 is: %s", base64_encoded_pkcs7);
-    PREPF(TLS, enroll_req->ecfg->tls->strerror);
-    tlsuv_certificate_t chain = NULL;
-    ziti_err = ZITI_PKCS7_ASN1_PARSING_FAILED;
-    TRY(TLS, enroll_req->ecfg->tls->parse_pkcs7_certs(
-            &chain, base64_encoded_pkcs7, strlen(base64_encoded_pkcs7)));
-    free(base64_encoded_pkcs7);
+int fetch_network_token(struct ziti_enroll_req *er) {
 
-    char *ca = NULL;
-    size_t total_pem_len = 0;
+    struct tlsuv_url_s url;
+    int rc = tlsuv_parse_url(&url, er->opts.url);
+    if (rc != 0) {
+        ZITI_LOG(ERROR, "URL[%s] is invalid", er->opts.url);
+        return ZITI_INVALID_CONFIG;
+    }
 
-    ziti_err = ZITI_INVALID_CONFIG;
-    TRY(TLS, chain->to_pem(chain, 1, &ca, &total_pem_len));
-    chain->free(chain);
+    er->tls = default_tls_context(NULL, 0);
 
-    ZITI_LOG(DEBUG, "CA PEM len = %zd", total_pem_len);
-    ZITI_LOG(TRACE, "CA PEM:\n%s", ca);
-
-    tls_context *tls = default_tls_context(ca, (strlen(ca) + 1));
-    if (enroll_req->ecfg->private_key != NULL) {
-        ziti_err = ZITI_KEY_LOAD_FAILED;
-        if (load_key_internal(tls, &enroll_req->ecfg->pk, enroll_req->ecfg->private_key) != 0) {
-            ZITI_LOG(WARN, "failed to load private key[%s]", enroll_req->ecfg->private_key);
-            if (enroll_req->ecfg->zej.method == ziti_enrollment_methods.ott &&
-                strncmp(enroll_req->ecfg->private_key, "pkcs11://", strlen("pkcs11://")) == 0) {
-                ZITI_LOG(INFO, "attempting to generate pkcs11 key");
-                TRY(TLS, gen_p11_key_internal(tls, &enroll_req->ecfg->pk, enroll_req->ecfg->private_key));
-            }
+    if (url.query != NULL) {
+        // TODO magic URL
+    } else {
+        // controller URL must be verifiable with default/OS cert bundle
```

Comment:
will there be a clear log message or error when trying to use network jwt without a verifiable os / cert bundle?